src/targeting/personalised.ts

```diff
@@ -0,0 +1,256 @@
+	 */
+	rdp: True | False | NotApplicable;
+};
+
+/* -- Methods -- */
+
+const getRawWithConsent = (key: string, state: ConsentState): string | null => {
+	if (state.tcfv2) {
+		if (state.tcfv2.consents['1']) return storage.local.getRaw(key);
+	}
+	if (state.ccpa) {
+		if (!state.ccpa.doNotSell) return storage.local.getRaw(key);
+	}
+	if (state.aus) {
+		if (state.aus.personalisedAdvertising) return storage.local.getRaw(key);
+	}
+
+	return null;
+};
+
+const getFrequencyValue = (
+	state: ConsentState,
+): PersonalisedTargeting['fr'] => {
+	const rawValue = getRawWithConsent('gu.alreadyVisited', state);
+	if (!rawValue) return '0'; // TODO: should we return `null` instead?
+
+	const visitCount: number = parseInt(rawValue, 10);
+
+	if (visitCount <= 5) {
+		return frequency[visitCount] ?? '0';
+	} else if (visitCount >= 6 && visitCount <= 9) {
+		return '6-9';
+	} else if (visitCount >= 10 && visitCount <= 15) {
+		return '10-15';
+	} else if (visitCount >= 16 && visitCount <= 19) {
+		return '16-19';
+	} else if (visitCount >= 20 && visitCount <= 29) {
+		return '20-29';
+	} else if (visitCount >= 30) {
+		return '30plus';
+	}
+
+	return '0';
+};
+
+const tcfv2AllPurposesConsented = (consents: TCFv2ConsentList) =>
+	Object.keys(consents).length > 0 && Object.values(consents).every(Boolean);
+
+const personalisedAdvertising = (state: ConsentState): boolean => {
+	if (state.tcfv2) return tcfv2AllPurposesConsented(state.tcfv2.consents);
+	if (state.ccpa) return !state.ccpa.doNotSell;
+	if (state.aus) return state.aus.personalisedAdvertising;
+
+	return false;
+};
+
+type CMPTargeting = Pick<
+	PersonalisedTargeting,
+	'cmp_interaction' | 'pa' | 'consent_tcfv2' | 'rdp'
+>;
+const getCMPTargeting = (state: ConsentState): CMPTargeting => {
+	if (state.tcfv2) {
+		return {
+			cmp_interaction: state.tcfv2.eventStatus,
+			pa: tcfv2AllPurposesConsented(state.tcfv2.consents) ? 't' : 'f',
+			consent_tcfv2: tcfv2AllPurposesConsented(state.tcfv2.consents)
+				? 't'
+				: 'f',
+			rdp: 'na',
+		};
+	}
+
+	if (state.ccpa) {
+		return {
+			consent_tcfv2: 'na',
+			rdp: state.ccpa.doNotSell ? 't' : 'f',
+			pa: state.ccpa.doNotSell ? 'f' : 't',
+		};
+	}
+
+	if (state.aus) {
+		return {
+			consent_tcfv2: 'na',
+			rdp: 'na',
+			pa: state.aus.personalisedAdvertising ? 't' : 'f',
+		};
+	}
+
+	return {
+		cmp_interaction: 'na',
+		consent_tcfv2: 'na',
+		rdp: 'na',
+		pa: 'f',
+	};
+};
+
+const isAdManagerGroup = (s: string | null): s is AdManagerGroup =>
+	adManagerGroups.some((g) => g === s);
+
+const getAdManagerGroup = (
+	state: ConsentState,
+): PersonalisedTargeting['amtgrp'] => {
+	if (!personalisedAdvertising(state)) {
+		storage.local.remove(AMTGRP_STORAGE_KEY);
+		return null;
+	}
+
+	const existingGroup = storage.local.getRaw(AMTGRP_STORAGE_KEY);
+
+	return isAdManagerGroup(existingGroup)
+		? existingGroup
+		: createAdManagerGroup();
+};
+
+const createAdManagerGroup = (): AdManagerGroup => {
```

Comment:
This is just personal taste but I'd define this function before it's used in `getAdManagerGroup`. If we wanted to enforce that I think we could add the [no-use-before-define](https://eslint.org/docs/rules/no-use-before-define) eslint rule.